### ISSUE TITLE
32-bit DLLInvoke Module Not Found

### ISSUE THREAD

**Issue by spurrme** [adamcaudill/libsodium-net#111]
When running a GenerateKeyPair call on a 32 bit Operating system I get the following message 

"Unable to load DLL 'libsodium.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)   at DynamicDllInvokeType.crypto_box_keypair(Byte[] , Byte[] ) at Sodium.PublicKeyBox.GenerateKeyPair()"

The libsdodium.dll and libsodium-64.dll are in the same directory that I am running the executable in.  When I first made the program I used an Any CPU build and ran it on a 64 bit system and everything worked perfectly (even with Mono).  I changed the build properties to x86 for both my program and the libsodium-net project, and that is when the above error started happening.  Not sure if this is an error in my understanding or a problem.

**Comment by adamcaudill** [comment]
> I changed the build properties to x86 for both my program and the **libsodium-net project**

`libsodium-net` will automatically use the proper libsodium library based on the process it's running under, so there shouldn't be a need to make any changes to this project. Did you get the same error before making that change?

**Comment by bitbeans** [comment]
Try to play with the Visual Studio Option: Prefer 32 bit

>If the Prefer32-bit check box is selected, the application runs as a 32-bit application on both 32-bit    and 64-bit versions of Windows. If the check box is cleared, the application runs as a 32-bit application on 32-bit versions of Windows and as a 64-bit application on 64-bit versions of Windows.

**Comment by spurrme** [comment]
Sorry guys I forgot to close this issue.  I used a process monitor to figure out what was going on.  I was missing a core dll for the x86 .NET framework.  It seems weird that the missing dll is not called libsodium.dll, but yet the error message states that.

**Comment by bitbeans** [comment]
Thank you for the feedback :)

**Comment by panodk** [comment]
@spurrme Hate to resurrect this thread but could you share what core dll was missing?  I'm encountering a similar error with the libsodium-64 even though the project works correctly in staging and dev environment.  

I have already set my PATH to the unmanaged directory containing libsodium.dll and libsodium-64.dll as instructed by this article: [How to include libsodium.net on ASP.NET] (http://stackoverflow.com/questions/28275944/how-to-include-libsodium-net-on-asp-net).

I am reviewing Process Monitor results but I have yet to find the causing factor.

Any help would be appreciated.  Thank you!

This simple test controller is failing in one environment:
```
		public string TestSodium()
		{
			string path = Environment.GetEnvironmentVariable("PATH");
			string binDir = System.IO.Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Bin");
			if (!path.Contains(binDir))
				Environment.SetEnvironmentVariable("PATH", binDir +";" + path);


			var t1 = Sodium.SecretKeyAuth.GenerateKey();
			string output = Sodium.Utilities.BinaryToHex(t1);
			return output;

		}
```

Related error:
```
[DllNotFoundException: Unable to load DLL 'libsodium-64.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)]
   DynamicDllInvokeType.sodium_init() +0

[TypeInitializationException: The type initializer for 'Sodium.SodiumCore' threw an exception.]
   Sodium.SodiumCore.GetRandomBytes(Int32 count) +0
```

**Comment by spurrme** [comment]
From what I remember of this issue, either the C++ or C# runtime needed to be installed/upgraded/repaired on the testing environment. If that doesn't help I will dig up my procmon dump tomorrow to get a more concrete answer.

**Comment by panodk** [comment]
I found it!  

Procmon identified the missing file as **msvcr100.dll**, the core library of the MS C++ runtime.  I downloaded [Microsoft Visual C++ 2010 SP1 Redistributable Package (x64)] (https://www.microsoft.com/en-us/download/details.aspx?id=13523), installed it, iisreset and then it began to work on my Windows 2012 server.  

Thanks again! :)

**Comment by cormacrelf** [comment]
This one just bit me, is there any reason to believe that the misreporting of the dll name is on libsodium-net's end?

**Comment by ScottAHoward** [comment]
This got me also.  I had to add this to my Global.asax
Hope you like VB.net as much as I do!
`        Dim path As String = Environment.GetEnvironmentVariable("PATH")
        Dim binDir As String = AppDomain.CurrentDomain.BaseDirectory & "Bin"
        Environment.SetEnvironmentVariable("PATH", path + ";" + binDir)`

**Comment by MuadDib** [comment]
I got this one as well. First part of the solution was adding path in Global.asax as few people already mentioned. 
But, I got the exact same error when deployed to our QA environment. After spending hours messing with permissions, trying to figure out if the dlls are in the right directory, and similar I ran into this post:
https://mspcontrol.org/forums/topic/add-user-error-unable-to-load-dll-libsodium-64-dll/#post-4765

> You have to install Visual C++ Redistributable for Visual Studio 2015 (13 Mb)
> https://www.microsoft.com/en-us/download/details.aspx?id=48145

In the end that solved the issue for me.
